### PR TITLE
unistd/file.c: open() on a directory with write access - set errno to EISDIR

### PR DESCRIPTION
**Pull request by Maxez**
## Description

Added POSIX compliant behaviour for open(), fopen() and freopen() on a directory.

## Motivation and Context

[JIRA: RTOS-132](https://jira.phoenix-rtos.com/browse/RTOS-132)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic, armv7a7-imx6ull.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: stdio/file.c [hunk 1]
```diff
@@ -146,9 +146,22 @@ static ssize_t safe_write(int fd, const void *buf, size_t size)
 
 static int safe_open(const char *path, int oflag, mode_t mode)
 {
+	struct stat s;
 	int err;
-	while ((err = open(path, oflag, mode)) < 0 && errno == EINTR)
+
+	if (oflag & (O_WRONLY | O_RDWR | O_APPEND)) {
```
Comment (nalajcie) [inline stdio/file.c:152]: `O_APPEND` should not be here, POSIX requires to always use one of the basic open flags (`O_APPEND` is an additional one)

**Review by nalajcie — COMMENTED** [review]
What about `open()`? POSIX requires for it to also return `EISDIR` error on trying to open dirs for writing.

image

So maybe this functionality should in fact be part of `open()` implementation?

**Comment by agkaminski** [comment]
@nalajcie You're right, it should be implemented in open(), it will take care of fopen() too

**Comment by Maxez** [comment]
I moved proposed checks to `open()` implementation - now `open()`, `fopen()` and `freopen()` should all correctly handle directories with write access.

**Review by agkaminski — APPROVED** [review]

File: stdio/file.c (hunk 1 above)
Comment (Maxez) [reply stdio/file.c]: Changed as requested. Please note that in that case we won't properly handle fopen() on an existing directory with "a" mode. The "a" mode gets translated to O_APPEND | O_CREAT (so doesn't pass our write access check) and we will successfully open directory in append mode.
I did check this case on Linux - it does set errno to EISDIR and return -1 in that case.

Comment (Maxez) [reply stdio/file.c]: Actually, shouldn't append mode ("a") [here](https://github.com/phoenix-rtos/libphoenix/blob/bd09e6dc360d76ef1b9facda9ecefa335aa5e58a/stdio/file.c#L83) be translated to O_APPEND | O_WRONLY | O_CREAT?
That would solve the issue.

Comment (Maxez) [reply stdio/file.c]: Added append mode flags in another [PR](https://github.com/phoenix-rtos/libphoenix/pull/146)

**Review by nalajcie — APPROVED** [review]
Seems valid and IMHO should support symlinks correctly.

I would welcome some unit test cases for these particularities (the path is already paved with `resolve_path` unit tests so it's easy to add).

**Pull request merged**